### PR TITLE
Authentication config file for in-memory backend

### PR DESCRIPTION
**Pull request by alexandre-merle**
"This PR provides a new config file format for the memory backend's authentication data, as well as a number of small utilities.

Edit: Please note that the checker writes the errors into the standard error output stream, in order for the user to see them. I am unsure whether this is the best solution or not, but I wanted to do that to ensure early failure if the configuration wasn't right. The thrown error in this case might deserve improvement also. I'm open to suggestions on that side :)"

Reopening of https://github.com/scality/S3/pull/8

### CODE REVIEW

**Comment by alexandre-merle** [comment]
@ironman-machine give me your status about that :)

File: lib/auth/in_memory/checker.js [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+        // eslint-disable-next-line no-param-reassign
+        data.arns[userObj.arn] = incr(data.arns[userObj.arn]);
+    }
+    if (checkExists(data, 'User', 'email', 'string', userObj)) {
+        // eslint-disable-next-line no-param-reassign
+        data.emails[userObj.email] = incr(data.emails[userObj.email]);
+    }
+    checkExists(data, 'User', 'keys', 'array', userObj);
+
+    if (userObj.keys) {
```
Comment (MathieuCassagne) [inline lib/auth/in_memory/checker.js]: Can't you do just like:
```
if (checkExists(data, 'User', 'keys', 'array', userObj))
    userObj.keys.forEach(keyObj => {
        // eslint-disable-next-line no-param-reassign
        data.keys[keyObj.access] = incr(data.keys[keyObj.access]);
    });
}
```

File: lib/auth/in_memory/checker.js [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+            incr(data.canonicalIds[accountObj.canonicalID]);
+    }
+
+    if (accountObj.users) {
+        if (checkType(data, 'Account', 'users', 'array', accountObj)) {
+            accountObj.users.forEach(userObj => checkUser(data, userObj));
+        }
+    }
+
+    if (accountObj.keys) {
```
Comment (MathieuCassagne) [inline lib/auth/in_memory/checker.js]: Same you can use ```checkExists()``` here.

Comment (alexandre-merle) [inline lib/auth/in_memory/checker.js]: Not this time, it's an optional field here :).

Comment (MathieuCassagne) [inline lib/auth/in_memory/checker.js]: Oh yes, my bad :)

**Comment by alexandre-merle** [comment]
@ironman-machine give me your status about that :)

File: lib/auth/in_memory/checker.js [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+function dumpErrors(checkData) {
+    let nerr = dumpCountError('CanonicalID', checkData.canonicalIds);
+    nerr += dumpCountError('Email', checkData.emails);
+    nerr += dumpCountError('ARN', checkData.arns);
+    nerr += dumpCountError('AccessKey', checkData.keys);
+
+    if (nerr > 0) {
+        return true;
+    }
+
```
Comment (LaurenSpiegel) [inline lib/auth/in_memory/checker.js]: If I'm reading this correctly, we log errors for duplicate items and then don't log any other formatting errors if there are duplicates?  Why?  Also, a few more comments would be helpful here.

File: lib/auth/in_memory/index.js [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+        };
+        this.usersBy = {
+            accessKey: {},
+            email: {},
+        };
+
+        /*
+         * This may happen if the file backend is not configured for S3.
+         * As such, we're managing the error here to avoid screwing up there.
+         */
```
Comment (LaurenSpiegel) [inline lib/auth/in_memory/index.js]: ?

File: lib/auth/in_memory/checker.js [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+function incr(count) {
+    if (count !== undefined) {
+        return count + 1;
+    }
+    return 1;
+}
+
+/**
+ * @param {object} data - the error collector object
```
Comment (rahulreddy) [inline lib/auth/in_memory/checker.js]: Please add function description here.

File: lib/auth/in_memory/checker.js [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+ * @param {string} container - the name of the entity that contains
+ *                             what we're checking
+ * @param {string} name - the name of the entity we're checking for
+ * @param {string} type - expected typename of the entity we're checking
+ * @param {object} obj - the object we're checking the fields of
+ * @return {boolean} true if the field exists and type is Ok
+ *                   false if an error was found and reported
+ */
+function checkExists(data, container, name, type, obj) {
```
Comment (rahulreddy) [inline lib/auth/in_memory/checker.js]: Please add function description here.

File: lib/auth/in_memory/index.js [hunk 7]
```diff
@@ -0,0 +1,145 @@
+/**
+ * Class that provides an internal indexing over the simple data provided by
+ * the authentication configuration file for the memory backend. This allows
+ * accessing the different authentication entities through various types of
+ * keys.
+ *
+ * @class Index
+ */
+export default class Index {
```
Comment (rahulreddy) [inline lib/auth/in_memory/index.js]: `Indexer` would be more appropriate here, as conventionally `index.js` just means root file for the project.

**Comment by michael-zapata** [comment]
Don't we want this in scality/Arsenal instead?

File: lib/auth/in_memory/checker.js [hunk 8]
```diff
@@ -0,0 +1,176 @@
...
+        arns: [],
+        canonicalIds: [],
+        keys: [],
+    };
+
+    if (authdata.accounts === undefined) {
+        checkData.errors.push({
+            txt: 'no "accounts" array defined in Auth config',
+        });
+        return dumpErrors(checkData);
```
Comment (ThibaultRiviere) [inline lib/auth/in_memory/checker.js]: i don't understand, why check here?

**Comment by alexandre-merle** [comment]
@ironman-machine and again :)

**Comment by DavidPineauScality** [comment]
End-to-End tests passed on https://github.com/scality/Integration/pull/74 both before this PR (http://ci.ironmann.io/gh/scality/Integration/1881) and after (http://ci.ironmann.io/gh/scality/Integration/1882)

Please review this so that we can FINALLY merge it...

**Comment by alexandre-merle** [comment]
Agreed with @DavidPineauScality , this pull request is open since 1 month now :)

**Comment by LaurenSpiegel** [comment]
There are open comments...

File: lib/auth/in_memory/checker.js (hunk 8 above)
Comment (DavidPineauScality) [inline lib/auth/in_memory/checker.js]: Sorry, I'm not sure I get the question, but my go-to answer is: There are no accounts to browse anyways, might as well return early.

**Comment by DavidPineauScality** [comment]
@LaurenSpiegel @rahulreddy updated :)

**Comment by DavidPineauScality** [comment]
Btw, https://github.com/scality/Integration/pull/74 need to be merged (and pushed to wip/ultron) first :)

**Comment by rahulreddy** [comment]
👍  Thanks @DavidPineauScality  and @alexandre-merle !

**Comment by LaurenSpiegel** [comment]
@ironman-machine just to be sure

**Comment by LaurenSpiegel** [comment]
End to end passed.  Merging.

**Pull request merged**